src/corona_hakab_model/clustered_matrix/clustered.py

```diff
@@ -0,0 +1,179 @@
+    __slots__ = ("size", "cluster_index", "local_indices", "clusters")
+
+    def __init__(self, clusters: Iterable[Collection[int]]):
+        self.size = sum(len(c) for c in clusters)
+        self.cluster_index = np.empty(self.size, dtype=int)
+        self.local_indices = np.empty(self.size, dtype=int)
+        self.clusters = []
+        for i, c in enumerate(clusters):
+            self.clusters.append(Cluster(c))
+            for j, a in enumerate(c):
+                self.cluster_index[a] = i
+                self.local_indices[a] = j
+
+    def batch_set(self, row, columns, probs, vals):
+        c_i = self.cluster_index[row]
+        l_i = self.local_indices[row]
+        local_columns = self.local_indices[columns]
+        self.clusters[c_i].batch_set(l_i, local_columns, probs, vals)
+
+    def row_set_prob_coff(self, row, coff):
+        c_i = self.cluster_index[row]
+        l_i = self.local_indices[row]
+        self.clusters[c_i].row_set_prob_coff(l_i, coff)
+
+    def col_set_prob_coff(self, col, coff):
+        c_i = self.cluster_index[col]
+        l_i = self.local_indices[col]
+        self.clusters[c_i].col_set_prob_coff(l_i, coff)
+
+    def row_set_value_offset(self, row, offs):
+        c_i = self.cluster_index[row]
+        l_i = self.local_indices[row]
+        self.clusters[c_i].row_set_value_offset(l_i, offs)
+
+    def col_set_value_offset(self, col, offs):
+        c_i = self.cluster_index[col]
+        l_i = self.local_indices[col]
+        self.clusters[c_i].col_set_value_offset(l_i, offs)
+
+    def __getitem__(self, item):
+        i, j = item
+        c_i = self.cluster_index[i]
+        c_j = self.cluster_index[j]
+        if c_i != c_j:
+            return None
+        l_i = self.local_indices[i]
+        l_j = self.local_indices[j]
+        return self.clusters[c_i][l_i, l_j]
+
+    def manifest(self, sample=None):
+        if sample is None:
+            sample = generator.random(sum(c.size ** 2 for c in self.clusters), dtype=np.float32)
+        s_next = 0
+        ret = ManifestClusters(self)
+        for i, c in enumerate(self.clusters):
+            bunch_size = c.size ** 2
+            is_manifest = sample[s_next: s_next + bunch_size].reshape(c.size, -1) < c.actual_probs()
+            s_next += bunch_size
+            ret.inners[i][is_manifest] = c.actual_vals()[is_manifest]
+        return ret
+
+
+class ManifestClusters(ManifestBase):
+    __slots__ = ("original", "inners")
+
+    def __init__(self, original: ClusteredSparseMatrix):
+        self.original = original
+        self.inners = [np.zeros_like(c.vals) for c in self.original.clusters]
+
+    def I_POA(self, v: np.ndarray, magic):
+        ret = np.empty(self.original.size, dtype=np.float32)
+        for cluster, manifest in zip(self.original.clusters, self.inners):
+            zipped_v = v[cluster.indices]
+            ret[cluster.indices] = [prod(magic.operate(m, v) for (m, v) in zip(row, zipped_v)) for row in manifest]
```

Comment:
i think you should do 1 - prod(...).
you can only prod the probabilities to NOT get infected in each of the meetings (since proding the probabilities of getting infected in each of the meeting would give you the probability you were infected a lot of times).
this way the prod would give you the probability that you weren't infected in any of the meetings, and 1 - prod is the probability that you were infected (at least in one meeting).

you should also make sure magic operator gives you the probability of NOT beeing infected in a single meeting, or otherwise is should be: 1 - prod(1 - magice.operate(m,v))

this applies for the cpp code as well.